### ISSUE TITLE
okta_rate_limiting resource broken in latest Okta Preview release

### ISSUE THREAD

**Issue by lucascantor** [okta/terraform-provider-okta#1099]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

- `0.14`

### Affected Resource(s)

- `okta_rate_limiting`

### Terraform Configuration Files

```hcl
resource "okta_rate_limiting" "rate_limiting" {
  login                  = "ENFORCE"
  authorize              = "ENFORCE"
  communications_enabled = true
}
```

### Debug Output

- https://gist.github.com/lucascantor/75a92c26598b3e56e7628224839ad30e

### Panic Output

- N/A

### Expected Behavior

- Terraform should generate a plan

### Actual Behavior

- `Error: failed to get client-based rate limiting: <nil>`

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

- N/A

### References

- N/A

**Comment by monde** [comment]
@lucascantor I'll have to discuss this one with the team if it will be fixed. It is built upon a private/internal API. I'm not sure looking at the issues leading up to the release the justification to do so. As a rule, we try to only interact with Okta through the public APIs that are documented. As it seems you have discovered an internal API changing without notice.

- https://github.com/okta/terraform-provider-okta/blame/master/okta/resource_okta_rate_limiting.go#L72-L82
- https://github.com/okta/terraform-provider-okta/blob/91fe4a59604bd0c3963b9202890792c637c5f987/sdk/rate_limit.go#L24-L36
- `url := "/api/v1/internal/rateLimits/clientRateLimitMode"`
- https://github.com/okta/terraform-provider-okta/pull/803
- https://github.com/okta/terraform-provider-okta/issues/800
- https://github.com/okta/terraform-provider-okta/issues/801

**Comment by lucascantor** [comment]
Thanks @monde 

I suppose in the meantime, our only option is to remove `okta_rate_limiting` resources from Terraform state?

**Comment by monde** [comment]
@lucascantor do you think you could do a PR to fix it? Typically each resource has an acceptance test but okta_rate_limit does not. I see a `okta/resource_okta_rate_limiting.go` it should also have a `okta/resource_okta_rate_limiting_test.go` file.

**Comment by lucascantor** [comment]
Would a test fix this for `okta_rate_limiting` resources already in Terraform state though?

I wasn't even able to remove ours, and instead decided to run `terraform state rm okta_rate_limiting.rate_limiting` manually.

Regardless, I'm so sorry, but I'm unfortunately completely unfamiliar with go, and wouldn't have the knowledge or time to work on a PR to for a test for this right now.

**Comment by monde** [comment]
@lucascantor I'll see if this feature still exists in the UI and if I can tweak the resource to have the previous behavior. If the endpoint is gone or its behavior has changed substantially the resource may have become obsolete. I'll post status when I get it.

**Comment by monde** [comment]
@lucascantor even though the endpoint is private, and could switch out without notice, it still works for me. Added an acceptance test in this PR #1121 . Ran it against a production org and a preview org.
```
$ TF_ACC=1 go test -tags unit -mod=readonly -test.v -run ^TestAccOktaRateLimiting_crud$ ./okta 2>&1
=== RUN   TestAccOktaRateLimiting_crud
--- PASS: TestAccOktaRateLimiting_crud (4.90s)
PASS
ok      github.com/okta/terraform-provider-okta/okta    5.259s
```
I also went and looked at Okta's source that manages this part the API. There are some guards on the code to ensure the API key has permission to manage rate limits on POST and read rate limits on GET. Perhaps the permissions on your org or those associated with your API token have changed?

**Comment by monde** [comment]
```
$ TF_LOG=1 TF_ACC=1 go test -tags unit -mod=readonly -test.v -run ^TestAccOktaRateLimiting_crud$ ./okta 2>&1 | tee /tmp/TestAccOktaRateLimiting_crud.log

$ grep "^\(\(PUT\|GET\|POST\|DELETE\) \/api\/\(v1\|internal\)\/\)\|\(HTTP\/2.0\)" /tmp/TestAccOktaRateLimiting_crud.log
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
POST /api/v1/internal/rateLimits/clientRateLimitMode HTTP/1.1
HTTP/2.0 200 OK
POST /api/internal/orgSettings/rateLimitNotificationSetting HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/internal/rateLimits/clientRateLimitMode HTTP/1.1
HTTP/2.0 200 OK
GET /api/internal/orgSettings/rateLimitNotificationSetting HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
GET /api/v1/users/me HTTP/1.1
HTTP/2.0 200 OK
```

**Comment by lucascantor** [comment]
@monde thank you. Our API key's permissions are definitely unchanged, but maybe this due to our org being in OIE or in a certain cell?

Either way, if you can't reproduce the issue, then I'm happy to close it and revisiting this later, as the `okta_rate_limiting` resource certainly isn't the most critical for us to manage via Terraform.

**Comment by monde** [comment]
@lucascantor the only thing that changed is your org was updated to OIE? If you run apply with TF_LOG=1 on can you show me the POST and response bodies again? I wonder if the response body is empty or at least missing `granularModeSettings`

**Comment by lucascantor** [comment]
@monde our org was updated to OIE, but to be clear, that happened a long time ago, well before we even started using the Okta Terraform provider at all.

Regardless, I already ran `terraform state rm okta_rate_limiting.rate_limiting` to remove this resource from Terraform state manually the other day.

Planning to re-create the resource succeeds, FWIW, but I unfortunately already deleted the original debug logs with the error(s) in them.
